### ISSUE TITLE
Investigate CDROM interaction with DOS Navigator

### ISSUE THREAD

**Issue by krcroft** [dreamer/dosbox-staging#158]
> In staging, there are some regressions: it can't mount .img files anymore (see topic on vogons). At least what i find from fast tests, maybe something else.

_Originally posted by @kas1e in https://github.com/dreamer/dosbox-staging/issues/144#issuecomment-581109417_

**Comment by krcroft** [comment]
CDROM image mount results, where mounts include:
- non ISO-9660, pure binary image - `imgmount d file.img -t iso`
- non ISO-9660, binary + cue - `imgmount d file.cue -t iso`
- ISO-9660 - `imgmount d file.iso -t iso`
- ISO-9660 plus audio tracks, with cue `imgmount d file.cue -t iso`

| Version/Date  | x86_64 | Amiga4 PPC32  | Linux PPC32  |
|---|---|---|---|
| Staging 2020-Feb-02 | ☑️  | ❌   | ☑️  |
| Staging 2020-Oct-1 | ☑️  | ❔   | ☑️  |
| SVN r4313-nopatch | ☑️  | ❔   | ☑️  |
| SVN r4293-nopatch  | ❔ | ❔  | ❔  |
| SVN r4293+PPC-JIT |  ❔ | ☑️  | ❔  |

Note: OS is not specified for x86_64 as image mounting has not been reported to differ between common OSes such as Linux, macOS, and Windows.

**Comment by krcroft** [comment]
Ok, changing gear nows. I've started focussing on MK3 specifically, and this where bug differences started to jump out:
- All of the above binaries failed to start MK3 when the `.bin` file was mounted (without `.cue`)
- When mounted with the `.cue` all SVN builds plus the 2019-Oct builds of staging started MK3.
 
@kas1e, please try the current SVN r4313 (with and without dynrec; doesn't matter) using `imgmount d mk3.cue -t iso`.  This should work for you.

I will work on bisecting this.

**Comment by dreamer** [comment]
Just to be sure, I tested my GOG copy of MK3 to check if I can reproduce the issue… and I can! This came as a surprise because I was testing MK3 in the past and it worked, and I don't have image mounting issues in other games.

I'll try to bisect the issue tomorrow.

GOG copy loads CD via .cue and image stored in MK3.GOG file, sha1sums: [mk3-image-sha1sum.txt](https://github.com/dreamer/dosbox-staging/files/4146258/mk3-image-sha1sum.txt)

**Comment by krcroft** [comment]
Thanks for helping @dreamer ! 
The auto-bisect just finished and landed on this commit -
```
commit eaeb001b1709ea88d51179a53423fffdfdc81c37
Author: krcroft <krcroft@users.noreply.github.com>
Date:   Wed Nov 20 00:04:14 2019 -0800

    Play into subsequent track(s) if playback length exceeds the the current track
    
    Issue reported by Dagar and Pr3tty F1y, and confirmed as a bug by ripsaw8080.
    Thank you!
    
    This fixes the GoG release of Betrayal at Krondor which (either due to CD mastering
    issues or a faulty rip), requests playback of a given track at the tail end
    of the prior track.
    
    In debugging and performing this fix, many debug messages were improved as well
    as making some small small code adjustments, such as using iterators to point to
    individual tracks (track->attribute) instead of using the tracks array
    (tracks[track -1].attribute).
```

It looks like the commit make functional and clean-up related changes; so will start to tease these appart.  Hopefully give us something to work with!

**Comment by krcroft** [comment]
@kas1e, can you test the `kc/cdda-last-track-1` branch with MK3?

(and ensure the CD is mounted using the `.cue`, as we've established that this is necessary for all versions of SVN and the earlier working versions of staging).

**Comment by kas1e** [comment]
@krcroft 
With ```kc/cdda-last-track-1 ``` MK3 start to works on PPC as well. At least when I mount it as a .CUE.

But issue remain that I can't mount pure .iso/.img files. I.e. mounting mk3 via mk3.cue, make it all works. But just "imgmount d mk3.img -t iso", says "MSCDEX: Failure: File is either no ISO/CUE image or contains errors". 

I will try not to build without opus as well.

**Comment by kas1e** [comment]
Interestingly, that when i mount MK3 via .CUE, and it works fine in game (i heard music), i can't play that music via usuall CD-Player in DosNavigator. 

But firstly will try to build without Opus, maybe something with it

**Comment by krcroft** [comment]
Hmm.. interesting! Yes; it could be something like that.  

Would you be able to send me your img and cue? 

Also, if you can enable debug statements in cdrom_image.cpp, it will provide a lot more information.  You can do this by commenting in the #define DEBUG 1 line near to top of the file, and then recompiling. 

When that's done, start mk3, and then save your console output to text file. Thanks @kas1e !

**Comment by kas1e** [comment]
Enabled debug, and made 3 tests:

1). Start dosbox, mount as CUE, start a game, start first level (so music plays from CDDA), and exit from game:

http://kas1e.mikendezign.com/aos4/dosbox/CD/when_mount_as_CUE_and_start_play_exit_mk3.txt

2). Start dosbox, mount as CUE, and tried to play via DOSNavigator's CD-Player:

http://kas1e.mikendezign.com/aos4/dosbox/CD/when_mount_as_CUE_and_try_cdplayer.txt

3). Start dosbox, and tried to mount C&C disk1.iso:

http://kas1e.mikendezign.com/aos4/dosbox/CD/when_tried_to_mount_an_iso.txt
And there is my mk3.img/mk3.cue:

And there my mk3.img/mk3.cue i use :

https://dropmefiles.com/Rug1E

**Comment by kas1e** [comment]
Found interesting moment.

If i replace in cdrom_image.cpp that:

```
int CDROM_Interface_Image::BinaryFile::getLength()
{
	// Guard: only proceed with a valid file
	if (file == nullptr) return -1;

	std::streampos original_pos = file->tellg();
	file->seekg(0, ios::end);
	const int length = static_cast<int>(file->tellg());
	file->seekg(original_pos, ios::end);
	return length;
}
```

On original version:

```
int CDROM_Interface_Image::BinaryFile::getLength()
{
	file->seekg(0, ios::end);
	int length = (int)file->tellg();
	if (file->fail()) return -1;
	return length;
} 
```

Then I can mount ISO,IMG,BIN files ! Retested many times.

But that didn't fix for me issue with CD-Player not being able to play CDTracks of MK3 when mount as a CUE.

Maybe something broken on "std::" side on our part (like maybe something in our newlib, etc). Will check more.

**Comment by kas1e** [comment]
What i also note, is that seems all mounted isos via CUE, have some issues when played via CD-Player. They didn't not starts as in original dosbox from track 2, but instead from track 6 or 7, and then i can go back to track 2. 

When we mount CUE which use some .ogg files or something from disk, then there is no problems with that.

**Comment by krcroft** [comment]
Thank you for all the feedback and content @kas1e .  Feel free to take down the file link; I've confirmed my image behaves the same as yours, so nothing revealing there. 

Awesome digging. Regarding GetLength(), the changes in staging ensure that the existing read position is not clobbered, and instead is returned to it's original position; however it's possible we can revert to the original code if we determine this protection isn't needed.

Looking at the CDROM debug statements from DosNavigator, it somehow can't move past the initial CDROM queries; something about the state or values it gets back where it feels the CDROM isn't ready yet -- so it keeps querying every second.  But playing the game first is able to put the CDROM conditions into a state DN seems happy with.  

Can you try playing the MK3 audio tracks with ftp://ftp.cubic.org/pub/player/ocp260pre6.zip ? (cue mount) It's the Open Cubic Player.  When it starts, nagivate down to D:\. It recognizes track 1 as data, and lets me play all the other audio tracks; and reads their durations correctly. I can pause and resume.  Curious what you see.

**Comment by kas1e** [comment]
> Awesome digging. Regarding GetLength(), the changes in staging ensure that the existing read position is not clobbered, and instead is returned to its original position; however it's possible we can revert to the original code if we determine this protection (or continuity) isn't needed.

I more curious wtf :) I mean how that protection can be related to issues with mounting... Maybe need to test some little test case for std::streampos usage, maybe it just broken on my side.

> Can you try playing the MK3 audio tracks with ftp://ftp.cubic.org/pub/player/ocp260pre6.zip ? It's the Open Cubic Player. When it starts, navigate down to D:. For me, it recognizes track 1 as data and lets me the audio tracks. It reads their durations correctly and can pause and resume. Curious what you see.

Yes, for me also it fine. I just run DOSBox, then OCP, then go to D:, and have that:

http://kas1e.mikendezign.com/aos4/dosbox/CD/cp_000.png

Choice any .CDA file plays fine.

But found one issue maybe worth of mention: When any file is playing and I hit enter to return to the menu, and then navigate to the "install.dat" (just navigate, didn't press enter on it), it then gives me heavy hardcore noise sound. Not sure if its bug of OCP or something to check out on our side..

Anyway .. what all of this means, that it can be just the second issue we have there in terms of CD-Audio as if DosNavigator's CD-Player stop working, maybe something/somewhere still wrong on our side? 

There i upload DN i use: 
http://kas1e.mikendezign.com/aos4/dosbox/CD/dn.zip

Just unpack, run dn.com,  choice in the menu at top "Utilities" and there "CD Player". So, if you just mount mk3.cue, and run cd-player you can't choose anything. If you run the game itself (or OCP for example), then, after you run DN's CD-Player you can navigate a bit. To the up limit only from 10 till 15 tracks in a loop, but can go lower till 2. In other words like something completely broken in terms of tracks numbering, etc when it comes to DN's CD-Player.

**Comment by kas1e** [comment]
I think, there can be some general issue with DosNavigator's CD-Player in whole DosBox, just with Staging it start to be worse. What I mean, is that while in original DOSBox playing of mk3's CDDA tracks mounted via CUE works like it, it still can't play any track past 15. And it also after track 15, jump on track 10.

On Staging it completely broken as I said before and can't play anything at all when we mount cue and tried to play MK3 audio, but then any other audio also didn't play well, and didn't start with track #2 as should.

But then, while in original DosBox with CD-Player from DosNavigator we didn't have such issues, it still, can't play more than 15 tracks of MK3. Just like there only 15 tracks and that all. As well as after 15 starts not from 2, but from 10.

That, of course, can be just limitation of DosNavigator's CD-Player not being able to play more than 15 tracks, or maybe it just Dosbox issues which just start to be bigger in staging.

**Comment by krcroft** [comment]
Hi @kas1e ; a bit more progress today.  I'm stepping through the MSCDEX specification and fixing responses to be more compliant in corner-cases.

I'm only a little ways into it, however these seem to have slightly helped DN without any negative impact to the handful of CD-based games I've tested (however, will have to do a lot more before declare them OK).  This is all just a work-in-progress for now. 

I'm not sure why DN rolls over at track 15; but clicking the track number lets me enter in a specific track to play - and that correctly plays but DN's visual track number is reduced by some quantity (5 or more). So if you play track 33, the display with show 22.  If you play track 47, at the 30-second mark there's some backwards speech; maybe an easter :egg:

**Comment by kas1e** [comment]
@krcroft 
Yep, tested ```kc/cdda-last-track-1``` and indeed things improve! It's all now reacts the same as with original DosBox. I.e. I can play MK3 audio via cd-player in DN, all tracks starts from "2" now, and in other  cue/isos its the same.

Through, I from time to time catch a crash. When I play a lot with mount/umount and starting DN's cd-player. At the moment all I can say it's just crash in GetAudioSub(), but not very often. I will try to enable some more debugging to see on what line exactly it crashes.

Now, another question remains: is it bug in DN's CD-Player that it can't visuall more than 15 tracks or DosBox one? I tried another version of  DN : same issue.

**Comment by kas1e** [comment]
Also as I can see, some tracks sometimes just stop playing in DN's cd-player. Like it plays track 3 of heroes2, and then on someplace just stop, and in console, I have "not paused and stopped", instead of usual "not paused and played".

**Comment by krcroft** [comment]
My hunch is that the DN is doing the right thing and expects (and possibly assumes) that MSCDEX responses will strictly adhere to the specification; where-as games and the Cubic Player might be less sensitive.  My reasoning is, we're dealing with version 1.51 of DN from '99, giving them eight years to work through any physical/hardware CDROM nuances with wide-spread usage on physical DOS machines (however, DN's development pre-dates DOSBox). Where as Cubic Player was developed up to 2009, spanning many years where people might use it in DOSBox and give feedback to the author.

> from time to time catch a crash. <snip> I will try to enable some more debugging

Great! If you have `gdb`, you can run DOSBox inside it with `gdb /path/to/dosbox`. Once inside the GDB prompt, use the `run [ARGUMENTS]` command to launch dosbox with whatever [ARGUMENTS] you normally launch DN with.  Ie.. `run start.bat`, or `run .`, etc..  Now with DN running, try to reproduce the crash.  GDB will report the crash, and you will be back at the GDB prompt.  Then simply type `bt` (or `backtrace`) to see the function call history and line numbers.

But first, DOSBox will need to built with debug symbols and reduced optimizations, such as `CFLAGS="-g -Og" CXXFLAGS="-g -Og"` (or `-O1` if `-Og` is not supported).

Regarding the repeats.. that's also valuable information; I will keep digging on the compliance side. Thanks again @kas1e .

**Comment by kas1e** [comment]
In meantime tested with original DosBox how it reacts when I play in DN's CD-Player 3rd track of Heroes2 cd_audio: there it didn't stop, but with last ```kc/cdda-last-track-1``` it surely stops all the time when playing 9 seconds. Strange :)

Btw, while we at it: have you maybe any idea, what can be "fast-fixed" in the original DosBox's cd_image.cpp's SDL_Sound code, so it will play endian-aware and not distorted sound? I just may in the meantime release another amigaos4 version of DosBox with that fix while we didn't deal with new staging-code.

**Comment by kas1e** [comment]
@krcroft 
Tested also your  8 hours ago commit ```strictly confirm to MSCDEX state transitions```,  this time Heroes2's tracks start only from 3st one (2nd one skipped), and 3rd one still stops on 9th second.

**Comment by krcroft** [comment]
Thanks @kas1e ; glad you're seeing what I am too.  Tomorrow I will make a new branch with just the critical bits changed, and see if we can mirror baseline behavior.

**Comment by kas1e** [comment]
And now I think it's not only DN. I for now tried to play with OCP some test opus-cd-of-mk3 you send me before, and while DN's CD-Player even can't play anything from, in OCP it at least can play at the beginning. But then, it shows only the first 33 tracks. Then, while I play a few of them, I start to have some "generic error" coming from OCP. Then, exit, second run, and I have visibly only tracks from 33 to 45.

So seems all that jumping between tracks and other problems happens not only with DN's CD-Player :(

**Comment by dreamer** [comment]
Decided to test HOMM2 music (GOG version) using OCP 2.6.0pre6. Initially it works ok, but then I get "generic error" and in successive runs new tracks beyond 33 appear - this is a regression compared to master (which seems to work perfectly for me for HOMM2).

@krcroft do you want me to do testing or should I wait for review? I don't want to derail discussion if you're fully aware about regressions on the test branch.

**Comment by krcroft** [comment]
Thanks for check; no need to take time testing the current branch; it was a bunch of tests without regard to existing functionality; I'm polishing up fixes on a new branch and will drop this one.

**Comment by krcroft** [comment]
@kas1e , @dreamer  - `kc/ccda-fixes-2` is ready for testing.  Curious how it goes on your side!

**Comment by dreamer** [comment]
Some quick tests:
- MK3 works again :) 
- GOG version of MK3 comes with audio tracks 2-47; when playing them in OCP I see 2-33; but in log I see dosbox-staging is loading all of them when mounting .cue file.
- HoMM2 works great
- GOG version of HoMM2 comes with audio tracks 2-43; when playing them in OCP I see 2-33 (same story, I see them in log when mounting, missing from OCP)

In the evening I'll do some more testing, but as far as games are concerned I see no problems (and MK3 works again, so that's nice :)). This limit of 33 tracks in OCP is suspicious - I believe the maximum number of tracks on CD is 99?

**Comment by kas1e** [comment]
Oh, it looks like something common between issues with OCP and DN's CD-Player.  Just in DN's we for some reason can't have more than 15, and in OCP can't have more than 33. So chances are high that fix will be the same. I will test more today a bit more deeply.

**Comment by krcroft** [comment]
The track number in the MSCDEX spec is an unsigned 8bit, so I've fixed that in the latest commit, however that hasn't solved the roll-over at track 16 / 33.  

This had me thinking that we're off by 1-bit, so I tried flipping between unsigned and signed 8 bit track numbers - however the results were identical. 

Another approach - in `dos_mscdex.cpp`, see line 953:

``` c++
Bit8u attr,track,index; 
mscdex->GetSubChannelData(drive_unit,attr,track,index,rel,abs);
mem_writeb(buffer+1,attr);
mem_writeb(buffer+2,track);
```

If we bit-shift the track value before writing it back to the DOS program (below) , we can get DOS Navigator to actually show larger track numbers: it jumps to 8, 16, 32, .. (as you would expect).  But this shows that DN's side can atleast handle the roll over of the larger values.

``` c++
Bit8u attr,track,index; 
mscdex->GetSubChannelData(drive_unit,attr,track,index,rel,abs);
mem_writeb(buffer+1,attr);
track <<= 1;
mem_writeb(buffer+2,track);
```

I feel like the answer is something subtle in here.

**Comment by kas1e** [comment]
@All

Oh, I found some interesting issues, which I was always thinking are MK3 issue (and it happens with original DOSBox as well): try to run MK3 with CDDA tracks (not GOG ones, just usual CDDA from CUE). Choice hero and start a Round. Then exit from (die, or escape): now, you will back into the menu, but (!) MUSIC DIDN'T PLAYS there! But if you will choose "start game", and press "ESC" ln the hero-choice screen then music on the menu will be played.

I thinking before that its MK3 issue itself, but now, I think that it can be just the same problem. Like, some tracks didn't find/plays/inits when should, but reinits/plays in some conditions only. Just like we have with CD players now.

Maybe its just a game's issue,  but probably worth to note.

**Comment by krcroft** [comment]
Regarding track number bitiness: I tried shifts in either direction; but those broke.  Assigned track number as a signed int (`int8_t`) to the ASPI-specified unsigned char and vice-versa, but saw no changes. 

Because valid tracks range from 1-99, the last bit of the 8-bit type is unused (127 and greater), I tried setting that last bit high (+=128) however DN actually honored it but couldn't display three digits so the TUI was empty; however  it properly queried for tracks in the 200+ range (but my track range checks on the DOSBox side triggered and quit)

If you don't touch the forward/back navigation buttons, DN will _play_ beyond track 15 and correctly play track's 16, 17, and so on (each time correctly getting the greater track number through sector queries).  However its GUI still displays the "rolled-back" track numbers.  So under the hood, this produces a correct round-trip interaction. 

Contrast this with _clicking_ the GUI track-forward button: in this mode, the roll-back behavior also affect playback, where under the hood DN queries for the rolled-back track number. 

Also check if any of this changes using a pure audio CD vs mixed/mode-2, but doesn't change anything.

**Comment by krcroft** [comment]
When left to play, DN's track display behaves like this:

| Display | Actual | Difference |
|--:|--:| --:|
| 1 |  1 | 0 |
|  2 | 2  | 0 |
| 3  | 3  | 0 |
| ...  | ...  | ... |
| 14  | 14  | 0 |
| 15  | 15  | 0 |
| 10  | 16  | 6 |
| 11  | 17  | 6 |
| 12  | 18  | 6 |
| ...  | ...  | ... |
| 23  | 29  | 6 |
| 24  | 30  | 6 |
| 25  | 31  | 6 |
| 20  | 32  | 12 |
| 21  | 33  | 12 |
| 22  | 24  | 12 |
| ...  | ...  | ... |

We can "correct" DN's track value by adding a correction factor of `(track / 16) * 6`, using integer math. For example:
- track 15: `(15 / 16) * 6` is zero, so we add nothing.
- track 16: `(16 / 16) * 6` is 6, bringing DN's display value of 10 back up to 16.
- track 31: `(31 / 16) * 6` is 6, bringing DN's display value of 25 back up to 31
- track 32: `(32 / 16) * 6` is 12, bringing DN's display value of 20 back up to 32

So for fun, we can change line 408 in `cdrom_image.cpp`:

``` c++
track_num = track->number + (track->number / 16) * 6;
```
However, this causes the underlying track being played to be shifted ahead by the correct factor.  So for example, DN might display track 16, but the underly track being played is 22.

The only way to really fix this would be to factor it into every API that involves track number submission or lookup.  However, there's nothing in the [MSCDEX audio spec](https://gist.github.com/abrasive/7a615e6dde0c1da962f9930cc63ee43d#file-mscdex-txt-L651) that mentioned this behavior. 

Destruction Derby is a good game to check this. It's CD has 20 tracks, and the game include a simple CD player in the option menu (and it displays the track number).  So in theory, we would see similar roll-back on track 16 - but we don't.  We can step the tracks all the way through 20, and it plays and displays each one correctly.  If you turn `DEBUG 1` on in cdrom_image.cpp, you'll also see it makes the same calls and queries than DN does.

Based on this, I have to conclude that DN's CD Player is buggy in the display & user navigation features. Regarding the Cubic player, it would be  interesting to see if the author could investigate why the player only displays 33 tracks.  In both cases, it would be interesting to hear what happens on real DOS and hardware.

**Comment by kas1e** [comment]
Yeah, it seems hardcore...

In other words, in the meantime, we can keep it as it, with broken dn and ocp players in the hope someone can check it on real hw? 

And that issue in MK3 I mention unrelated and just a game's bug? I mean that one: https://github.com/dreamer/dosbox-staging/issues/158#issuecomment-582991013

**Comment by krcroft** [comment]
Oh, yes, I suspect that's a bug in mk3, given other games perform the same interactions with mscdex and the cdrom code (and can repeat previously played tracks, stop, resume, and so on; over and over in the same run of the game).  

For now, this is as far as I can take it.  That is, until your next interesting discovery :-)

**Comment by dreamer** [comment]
Fix is now merged; do we want to keep this issue open for investigating audio track limits in OCP and DOS Navigator?

**Comment by krcroft** [comment]
Thanks @dreamer ; I think it makes sense to close it given mpxplay works flawlessly and is making the same mscdex API calls, and all games tested (including those with more than 35+ tracks) also work as expected.  So I believe we're simply seeing shortcomings in these two specific packages.

**Comment by dreamer** [comment]
I mentioned in the review, that mpxplay does not work for me, but perhaps I am doing something weird - but mounted .cue files work great in games and in OCP now, so I think mpxplay is not a critical issue.

**Comment by kas1e** [comment]
@krcroft 
Found another regression in terms of new CD_Rom code: When I mount simple .cue file, with just one single .ogg in it, then, when I play it from DN's CD-Player , the "seconds" is not counted at all. Everything stays at 0:00. 

While with original DosBox, when I mount the same test, then I can see how in DN's CD-Player seconds counts correctly.

Another issue report one of the users who test my test-build with this code, he says that for him HMMMs ogg didn't work, it mounts, but then errors:

```
DROM: Loaded homm2_01.ogg [44100 Hz, 2-channel, 1.9 minutes]
CDROM: Loaded homm2_02.ogg [44100 Hz, 2-channel, 2.0 minutes]
CDROM: Loaded homm2_03.ogg [44100 Hz, 2-channel, 1.8 minutes]
CDROM: Loaded homm2_04.ogg [44100 Hz, 2-channel, 2.8 minutes]
CDROM: Loaded homm2_05.ogg [44100 Hz, 2-channel, 2.6 minutes]
CDROM: Loaded homm2_06.ogg [44100 Hz, 2-channel, 2.6 minutes]
CDROM: Loaded homm2_07.ogg [44100 Hz, 2-channel, 2.4 minutes]
CDROM: Loaded homm2_08.ogg [44100 Hz, 2-channel, 2.5 minutes]
CDROM: Loaded homm2_09.ogg [44100 Hz, 2-channel, 2.9 minutes]
CDROM: Loaded homm2_10.ogg [44100 Hz, 2-channel, 2.0 minutes]
CDROM: Loaded homm2_11.ogg [44100 Hz, 2-channel, 2.1 minutes]
CDROM: Loaded homm2_12.ogg [44100 Hz, 2-channel, 1.9 minutes]
CDROM: Loaded homm2_13.ogg [44100 Hz, 2-channel, 2.1 minutes]
CDROM: Loaded homm2_14.ogg [44100 Hz, 2-channel, 2.5 minutes]
CDROM: Loaded homm2_15.ogg [44100 Hz, 2-channel, 1.8 minutes]
CDROM: Loaded homm2_16.ogg [44100 Hz, 2-channel, 2.0 minutes]
CDROM: Loaded homm2_17.ogg [44100 Hz, 2-channel, 2.2 minutes]
CDROM: Loaded homm2_18.ogg [44100 Hz, 2-channel, 0.4 minutes]
CDROM: Loaded homm2_19.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_20.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_21.ogg [44100 Hz, 2-channel, 1.0 minutes]
CDROM: Loaded homm2_22.ogg [44100 Hz, 2-channel, 0.2 minutes]
CDROM: Loaded homm2_23.ogg [44100 Hz, 2-channel, 0.9 minutes]
CDROM: Loaded homm2_24.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_25.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_26.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_27.ogg [44100 Hz, 2-channel, 1.1 minutes]
CDROM: Loaded homm2_28.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_29.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_30.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_31.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_32.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_33.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_34.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_35.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_36.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_37.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_38.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_39.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_40.ogg [44100 Hz, 2-channel, 0.1 minutes]
CDROM: Loaded homm2_41.ogg [44100 Hz, 2-channel, 1.3 minutes]
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
CDROM: GetTrack at sector 4294967149 is outside the playable range
```

**Comment by krcroft** [comment]
Thanks @kas1e ; will look into both (and reopened this issue)

**Comment by kas1e** [comment]
@krcroft 
Oh, did you see r4320 in the original dosbox ?

**Comment by dreamer** [comment]
c91f46905b99d709545c11deb6577856e8eac2f3 I will merge it in after I'll clean up r4319, but am interested if it solves the problem without introducing regressions in other players.

**Comment by kas1e** [comment]
I tested now, and it surely solves issues with dos-navigator's CD-Player. I surely can play all the MK3 tracks till 47, without that "-5" tracks jump, and can go past 15 for sure.

But it didn't fix the issue with OCP, where it limited to 33 still.

I also do check if regression happens: at least not with MK3, C&C, and Rayman.

**Comment by krcroft** [comment]
I couldn't figure it out in my comment here - https://github.com/dreamer/dosbox-staging/issues/158#issuecomment-582978042 

Ripsaw8080 to the rescue - thank you!  

My guess is the vast majority of games don't use this call; instead, they already know the track they want to play (plus any desired time offset); so that's why this might have gone so long undiscovered. 

I haven't had time to check it, but will.

**Comment by kas1e** [comment]
Interesting is with OCP it's the same kind of bug somewhere in DOSBox, or its OCP itself?

**Comment by krcroft** [comment]
OCP makes the same queries as Dos Navigator, so my guess is there's a bug in OCP.  I wouldn't doubt if ripsaw is working on a patch or runtime TSR to fix that too :-)

**Comment by kas1e** [comment]
About that previous report when one of testers report those "out of range" errors when play ogg files of hmm2 : it was his fauilt. 

He mount as disk D iso from Hmm1 , and then mount as disk E Hmm2. Then , when he run game it probabaly tries to get tracks from disk D and fail because of wrong tracks.

But then, question arise: is there bug or not in our code in that terms. I mean, if he install game from D, and then mount it as E, then of course it will try to get tracks from D, that pure bug on his side.

But if he install it from E, then mount D as something else, and dosbox code trying to play tracks from D, but not from E , then its bug in our code. Need to recheck it.

**Comment by krcroft** [comment]
Thanks for following up and sorting out that PEBKAC issue with out of bounds reads. I remember reading about how some games assumed D: has the audio tracks, regardless of where the data files were installed from.  This is because most users had a single soundcard with a single analog CD audio header plug.. therefore even if a user had two CD drives, there were likely to only have one capable of feeding the audio to their soundcard (and that was almost always the first drive).

That said; I'm not sure if HOMM2 assumes that or not.. or if it records the CD drive unit number used during the install.

**Comment by kas1e** [comment]
That what we find:

The sound works if I mount HoM&M 2 as one ISO no matter if the letter D or E and even as F. If I mount more ISO images, e.g. HoM&M I and HoM&M II, the sound works if HoM&M II is mounted first, i.e. D. If I mount it as a second image, i.e. Hom & M I is as D and HM&M 2 on E, then the sound does not work. The conclusion is that a game or DOSBox always tries to search for music with the highest letter to which any ISO image is assigned.

**Comment by kas1e** [comment]
Oh, and i found what wrong with "seconds" counting in DN :

Its first track which didn't counts. I mean, do just that:

```
FILE "test.ogg" MP3
    TRACK 01 AUDIO
        INDEX 01 00:00:00
```

Then mount, run DN's CD-Player : seconds didn't count, while music plays.

Now, do that:

```
FILE "test.ogg" MP3
    TRACK 01 AUDIO
        INDEX 01 00:00:00
FILE "test2.ogg" MP3
    TRACK 02 AUDIO
        INDEX 01 00:00:00
```

(test2.ogg it just copy of test.ogg)

And now run DN's CD-player: first track plays fine, but _no_ second counting happens. But in the second track all fine.

Can you reproduce that on your side too ?

**Comment by dreamer** [comment]
I know that .CUE file format is not formally specified and was somehow convinced, that audio CDs require a minimum of 2 tracks (with one track possibly "hidden") - while a single track cue would work only for CDs with binary data; there are other such limitations related to specifics of Red Book (such as minimum length of track being 4s or only 2 indexes per track).

However I can't find it anywhere, and applications tend to go beyond limitations of .cue / CD audio formats (e.g. we already allow 48kHz tracks, which are not possible on physical CD due to historical reasons, foobar2000 allows for more than 99 tracks, etc).

The closest I could find is: [link](https://web.archive.org/web/20160201021136/http://digitalx.org/cue-sheet/syntax/#track), but it only says "minimum of 1 track per file".

@krcroft you know Red Book format somewhat - are there any real limitations preventing us from handling single-track files? We could always treat them as a special case if it contradicts some parts of algorithm for getting all the audio offsets just right.

**Comment by krcroft** [comment]
Yes; the one-audio-track time count issue is definitely a bug that kas1e found - haven't had time to dig into it on my side yet, but it's now up in my queue after getting the asset/log fetcher done.  

One audio-track or one data-track is acceptable - it will be abutted with the lead-out track, bringing the track count to two (making it a valid CD).

Although we do allow going beyond spec on the encoded file side, we always stay within spec once those samples land on the RedBook side (and we actually have more compliance checks than baseline dosbox). For example, baseline DOSBox allows sector start, sector length, and track numbers to all be their data-type's maximum value (4-billion, I believe as they're all 32-bit ints.. maybe bigger).

But no ripped phsical CDROM will ever be able to exceed 99 tracks, or request a sector position or length beyond 100 minutes (in redbook-sectors).  Because those were the physical limits of the CDROM; even 100-minute CDROMs were quite rare, but did come out the very late 90s. 

So applications and games that used real CDROMs will (or should) always behave in-spec, and we definitely want to catch those non-spec requests (like we saw with the HOMM2 CD-mixup; atleast some information is better than just quitting to a black screen and can tip off the user that something might be amiss -- perhaps we can improve the error reporting more, with tips of what to check?).

**Comment by krcroft** [comment]
@kas1e , @dreamer  just pushed `kc/fix-cdda-3`; I have an appt today but hopefully can test extensively tonight.  Hoping you can give it a whirl in the meantime.

**Comment by kas1e** [comment]
@krcroft 
Yes, it did the trick! Better and better!:)

**Comment by dreamer** [comment]
#181 is now merged-in.

**Comment by kas1e** [comment]
@krcroft 
Wait! We can't close this one ever! :)

I found a reproducible crash (at least on my side, on AmigaOS4/PPC). To reproduce, build the latest staging-code and then:

1). run DOSBox
2). ```imgmount d any.cue -t iso```
3). run DN's CD-Player, press play. Press stop. close CD-Player
4). ```mount -u d```
5). ```imgmount d any.cue -t iso```
6). run DN's CD-Player => crash.

any.cue can be just the same. not necessarily different.

On my side always reproducible, with any .cue files. Through, it may no crash from the very first time, but surely crash on second time. Most of the time just from the first test.

I build debug version of it, so to catch a crash, and that what it shows me:

```
Crash log for task "dosbox_staging"
Generated by GrimReaper 53.19
The crash occurred in module dosbox_staging at address 0x7E8188D0
Type of crash: DSI (Data Storage Interrupt) exception
Alert number: 0x80000003

Register dump:
GPR (General Purpose Registers):
   0: 7E80FE14 5F9FF3F0 00000002 5FA8DF1C 5F9FF465 5F9FF464 5F9FF463 5F9FF458
   8: 5F9FF460 00000005 5FAB3894 00001560 35955555 63F22C00 00000010 00000001
  16: 61B27B80 023C0000 67F9C9D0 5FB5B100 5FB5B0D8 00000000 023C0000 7E86FF00
  24: 6FFAB180 5F9FF463 5F9FF464 5FAB3840 5FA4E8A8 5F9FF465 5F9FF464 5EC68450

SPRs (Special Purpose Registers):
           Machine State (msr) : 0x0002F030
                Condition (cr) : 0x00000000
      Instruction Pointer (ip) : 0x7E8188D0
       Xtended Exception (xer) : 0x682E28E6
                   Count (ctr) : 0x682E28FA
                     Link (lr) : 0x021B29A4
            DSI Status (dsisr) : 0x5BBBBDC0
            Data Address (dar) : 0x01831E68


Symbol info:
Instruction pointer 0x7E8188D0 belongs to module "dosbox_staging" (PowerPC)
Symbol: _ZN21CDROM_Interface_Image11GetAudioSubERhS0_S0_R4TMSFS2_ + 0x70 in section 1 offset 0x000BE858

Stack trace:
    [cdrom_image.cpp:439] dosbox_staging:_ZN21CDROM_Interface_Image11GetAudioSubERhS0_S0_R4TMSFS2_()+0x70 (section 1 @ 0xBE858)
    [dos_mscdex.cpp:454] dosbox_staging:_ZN7CMscdex17GetSubChannelDataEhRhS0_S0_R4TMSFS2_()+0x8c (section 1 @ 0xB5D9C)
    [dos_mscdex.cpp:944] dosbox_staging:_ZL18MSCDEX_IOCTL_Inputjh()+0x400 (section 1 @ 0xB8094)
    [dos_mscdex.cpp:1052] dosbox_staging:_ZL24MSCDEX_Interrupt_Handlerv()+0x148 (section 1 @ 0xB8610)
    [dos_mscdex.cpp:807] dosbox_staging:_ZN7CMscdex17SendDriverRequestEtj()+0x94 (section 1 @ 0xB88C4)
    [dos_mscdex.cpp:1215] dosbox_staging:_ZL14MSCDEX_Handlerv()+0x3d4 (section 1 @ 0xB8CD4)
    [dos_misc.cpp:48] dosbox_staging:_ZL13INT2F_Handlerv()+0x44 (section 1 @ 0x9BF40)
    [dosbox.cpp:143] dosbox_staging:_ZL11Normal_Loopv()+0x68 (section 1 @ 0x2180)
    [dosbox.cpp:322] dosbox_staging:_Z17DOSBOX_RunMachinev()+0x24 (section 1 @ 0x2258)
    [callback.cpp:105] dosbox_staging:_Z19CALLBACK_RunRealInth()+0x54 (section 1 @ 0x97E8)
    [shell_misc.cpp:549] dosbox_staging:_ZN9DOS_Shell7ExecuteEPcS0_()+0x8a0 (section 1 @ 0x20823C)
    [shell_cmds.cpp:164] dosbox_staging:_ZN9DOS_Shell9DoCommandEPc()+0x244 (section 1 @ 0x202C88)
    [shell.cpp:293] dosbox_staging:_ZN9DOS_Shell9ParseLineEPc()+0x270 (section 1 @ 0x1FB588)
    [shell.cpp:376] dosbox_staging:_ZN9DOS_Shell3RunEv()+0x1f0 (section 1 @ 0x1FC880)
    [shell.cpp:788] dosbox_staging:_Z10SHELL_Initv()+0xa60 (section 1 @ 0x1FC354)
    [sdlmain.cpp:2685] dosbox_staging:main()+0x660 (section 1 @ 0x1164E8)
    native kernel module newlib.library.kmod+0x00001e8c
    native kernel module newlib.library.kmod+0x00002dc8
    native kernel module newlib.library.kmod+0x000030b8
    dosbox_staging:_start()+0x170 (section 1 @ 0x1920)
    native kernel module dos.library.kmod+0x0002a40c
    native kernel module kernel+0x0006eb80
    native kernel module kernel+0x0006ebc8

PPC disassembly:
 7e8188c8: 81230000   lwz               r9,0(r3)
 7e8188cc: 9101000c   stw               r8,12(r1)
*7e8188d0: 81290018   lwz               r9,24(r9)
 7e8188d4: 90e10008   stw               r7,8(r1)
 7e8188d8: 7d2903a6   mtctr             r9

System information:

CPU
 Model: Freescale P5020 (E5500 core) V1.2
 CPU speed: 1995 MHz
 FSB speed: 798 MHz
 Extensions:

Machine
 Machine name: AmigaOne X5000/20
 Memory: 2097152 KB
 Extensions: bus.pci bus.pcie


```

As can be seen its DSI (data storage interrupt) crash, and crashes on ```cdrom_image.cpp:439```, which are ```const uint32_t sample_rate = player.trackFile->getRate();``` from the ```CDROM_Interface_Image::GetAudioSub```.

Crashed instruction is "lwz r9,24(r9)" which means load data from r9 (which are 0x00000005) offset 24 which I not sure what is it all but can be some NULL-Pointer access maybe?  Some oses (like amigaos4 in our case) don't set uninitialized memory to null, so this isn't necessarily uninitialized memory and in this case, a pointer may really have been set to null. 

But I not sure if it NULL Pointer issues at all. Check plz if you have the same on your x86 config

**Comment by krcroft** [comment]
Another great find! 

When we load a new ISO for a given drive (say, D:\), its vector of `Track` objects (one for each track in the cue) is cleared and then repopulated with the new cue's tracks (line 857):

``` c++
bool CDROM_Interface_Image::LoadIsoFile(char* filename)
{
	tracks.clear();
        (then repopulated with new tracks)
```

Meanwhile, there's a single global object called `Player` that does the work of playing a single audio track (You'll see lines like `player.track`... lines), which points to the desired playback track within the above vector of `tracks`.  

When DOSBox is started up, the player's `.track` points to nothing until Playback starts, at which point it's assigned the track (line 571):

``` c++
bool CDROM_Interface_Image::PlayAudioSector(uint64_t start, uint64_t len)
{
   ...
	// Update our player with properties about this playback sequence
	player.cd = this;
	player.trackFile = trackFile;
	player.startSector = start;
	player.totalRedbookFrames = len;
  ...
```

When the DOS program asks the MCSDEX driver for the current playback position, it calls `GetAudioSub` function where you've discovered this new crash.  The function has a couple protections in place to prevent accessing invalid data.. first, the vector `tracks` need to be not-empty, and then the `player.trackFile` pointer needs to be valid:

``` c++
bool CDROM_Interface_Image::GetAudioSub( ... )
   ...
	if (!tracks.empty()) { 	// We have a useable CD; get a valid play-position
		track_iter track = tracks.begin();
		// the CD's current track is valid
		if (player.trackFile) {
			const uint32_t sample_rate = player.trackFile->getRate();
   ...
```

On first start, these checks work great if the DOS program queries playback -before- it attempts to play a track; because the `player.trackFile` will be empty and that if statement will fail. 

However, after unmounting and remounting a new cue **but-before-playback** the `player.trackFile` is left still pointing to it's prior track! But that track object is long long; blown away in the `tracks.clear()` call.

So when we empty the track vector, `tracks.clear()`, we need to similar to the player's `.trackFile` pointer to null - because when those tracks are cleared, all prior pointers  to them become invalid.

Likewise, when the umbrella CD object (that internally holds the vector of `tracks`) is destroyed (line 311, the destructor function `~CDROM_Interface_Image()`), the global `player` object is still alive and well.  So this is another boundary where tracks switch from valid to invalid, and we similarly should null-out the `player.trackFile` pointer.

I'll poke through and look for any other pointers that might be affected.

**Comment by krcroft** [comment]
Draft PR on branch `kc/fix-remount-null-1`, ready for confirmation.
Thanks again @kas1e !

**Comment by kas1e** [comment]
The crash is fixed, yes! Thanks krcroft for not ending the work you start on all this CD-based code :) 

Will keep testing all this heavily :)

**Comment by krcroft** [comment]
Tentatively closing this. 
@kas1e, if we find new issues, we can open a new issue, as this thread is now quite large and winding, plus it takes quite  bit of time to load and navigate on mobile.
